Assignment6.Rmd

```diff
@@ -237,12 +236,72 @@ umich_tree2_opt <- prune(tree = umich_tree2,
 
 # Plot optimal second tree model
 post(umich_tree2_opt, file = "umich_tree2_opt.ps", title = "Optimal Second model")
+
+### 3. Model selection: using Grid search ###
+# Establish a list of possible values for minsplit and maxdepth
+minsplit <- seq(1, 4, 1)
+maxdepth <- seq(1, 6, 1)
+
+# Create a data frame containing all combinations 
+hyper_grid <- expand.grid(minsplit = minsplit, maxdepth = maxdepth)
+
+# Number of potential models in the grid
+num_models <- nrow(hyper_grid)
+
+# Create an empty list to store models
+umich_models <- list()
+
+## Model training: train the grid of models
+for (i in 1:num_models) {
+
+    # Get minsplit, maxdepth values at row i
+    minsplit <- hyper_grid$minsplit[i]
+    maxdepth <- hyper_grid$maxdepth[i]
+
+    # Train a model and store in the list
+    umich_models[[i]] <- rpart(formula = GPAO ~ SUBJECT + GRD_PTS_PER_UNIT + TERM, 
+                               data = umich_valid, 
+                               method = "anova",
+                               minsplit = minsplit,
+                               maxdepth = maxdepth)
+}
+
+## Evaluate model performaces
+# Number of potential models in the grid
+num_models <- length(umich_models)
+
+# Create an empty vector to store RMSE values
+rmse_values <- c()
+
+# Compute validation RMSE
+for (i in 1:num_models) {
+
+    # Retrieve the i^th model from the list
+    model <- umich_models[[i]]
+    
+    # Generate predictions on grade_valid 
+    pred <- predict(object = model, newdata = umich_valid)
+    
+    # Compute validation RMSE
+    rmse_values[i] <- rmse(actual = umich_valid$GPAO, 
+                           predicted = pred)
+}
+
+
+# Identify the model with smallest validation set RMSE
+best_model2 <- umich_models[[which.min(rmse_values)]]
+
+# Compute test set RMSE on best_model
+umich_pred_valid <- predict(object = best_model2,
+                            newdata = umich_test)
+rmse(actual = umich_test$GPAO, 
+     predicted = umich_pred_valid)
```

Comment:
I appreciate your hard work to compare the two models by looking at different aspects of their performance! One thing you may want to pay attention is how to interpret the results and consider different measures as a whole. It would be great if you could write some interpretation and explain your decision on the model you choose to go with in this case :)